### ISSUE TITLE
Support for irregular observation intervals for area charts.

### ISSUE THREAD

**Issue by chrisjcameron** [bokeh/bokeh#2918]
I have not been able to find a way to specify the x-values of the data points used to construct the area plot (and the index argument seems to address a different issue). 

In the documentaiton example at http://bokeh.pydata.org/en/latest/docs/gallery/area_chart.html, it seems the index locations of the data points are used as the x value. 

Consider these other use cases:

1. The observations are actually taken at different time intervals and the user wants the x axis in terms of time points. 

2. The first observation has an x value of 14 instead of 0. 

3. The observations are collected at x values corresponding to a Fibonacci sequence.

There does not seem to be a documented way to handle these cases. Please consider updating the function or the documentation to reflect these use cases.

**Comment by rothnic** [comment]
@chrisjcameron: could you provide some sample data or a script to produce sample data? This chart has been reworked, but it is hard to know if we are covering the case you are describing.

**Comment by chrisjcameron** [comment]
If I understand the documentation correctly, I should be able to specify an index arg that provides the 
desired behavior. The following code will not produce a plot and complains about `IndexError: list index out of range`. The index argument works only when `index =range(14)`. Even simple shifting of the axis as in `index = [x+2 for x in range(14)]` produces errors. 


    # create some example data
    xyvalues = OrderedDict(
        python=[2, 3, 7, 5, 26, 221, 44, 233, 254, 265, 266, 267, 120, 111],
        pypy=[12, 33, 47, 15, 126, 121, 144, 233, 254, 225, 226, 267, 110, 130],
        jython=[22, 43, 10, 25, 26, 101, 114, 203, 194, 215, 201, 227, 139, 160],
    )

    #output_file(filename="area.html")

    area = Area(
        xyvalues, index = 'python',
        title="Responsive Area Chart",
        xlabel='time', ylabel='memory',
        stacked=True, legend="top_left",
        responsive=True
    )

**Comment by rothnic** [comment]
Here is the updated example: [area.py](https://github.com/bokeh/bokeh/blob/master/examples/charts/file/area.py). When provided like the old example, there is an implied index of the length of the data. I'm not sure about the old implementation, but maybe it is using the input index to access the values in the list, which would cause an index out of range in your case, but not completely sure.

What you'd want to do with the new charts is use pandas, which has a built in index, or include another list of values for the sample times as an index column. Then when you are creating your data, if there isn't a sample value, you should put NaN or 0. In the end, your data should have columns that are all the same length

I'd have to check what the behavior would be with nans vs 0, but that is what I'd expect to support. Just include whatever index you want in the data you input, then you'd have x='index', y=['python', 'pypy', 'jython'].

**Comment by rothnic** [comment]
@chrisjcameron 

Btw, this new implementation is going to be updated and more complete for 0.11, and there are some outstanding PRs, but this is what you'd do. There is a use of referencing the dataframe index that isn't currently working correctly, but this does:

```python
from bokeh.charts import Area, show, output_file

# create some example data
data = dict(samples=[x + 2 for x in range(14)],
            python=[2, 3, 7, 5, 26, 221, 44, 233, 254, 265, 266, 267, 120, 111],
            pypy=[12, 33, 47, 15, 126, 121, 144, 233, 254, 225, 226, 267, 110, 130],
            jython=[22, 43, 10, 25, 26, 101, 114, 203, 194, 215, 201, 227, 139, 160])

output_file(filename="area.html")

area1 = Area(data, title="Area Chart", x='samples', y=['python', 'pypy', 'jython'], 
             ylabel='memory', legend="top_left")

show(area1)
```

**Comment by chrisjcameron** [comment]
This is hopeful news, thanks for the update.

**Comment by rothnic** [comment]
Ok, going to close, but we can open another if that doesn't meet the full intent. Closed by #3024.
